src/gpuarray_buffer_cuda.c

```diff
@@ -304,61 +295,139 @@ static void cuda_deinit(void *c) {
   cuda_free_ctx((cuda_context *)c);
 }
 
-static gpudata *cuda_alloc(void *c, size_t size, void *data, int flags,
-			   int *ret) {
-    gpudata *res;
-    cuda_context *ctx = (cuda_context *)c;
-    int fl = CU_EVENT_DISABLE_TIMING;
+static void find_best(cuda_context *ctx, gpudata **best, gpudata **prev,
+                     size_t size) {
+  gpudata *temp, *tempPrev = NULL;
+  *best = NULL;
 
-    if ((flags & GA_BUFFER_INIT) && data == NULL) FAIL(NULL, GA_VALUE_ERROR);
-    if ((flags & (GA_BUFFER_READ_ONLY|GA_BUFFER_WRITE_ONLY)) ==
-	(GA_BUFFER_READ_ONLY|GA_BUFFER_WRITE_ONLY)) FAIL(NULL, GA_VALUE_ERROR);
+  for (temp = ctx->freeblocks; temp; temp = temp->next) {
+    if (temp->sz >= size && (!*best || temp->sz < (*best)->sz)) {
+      *best = temp;
+      *prev = tempPrev;
+    }
+    tempPrev = temp;
+  }
+}
 
-    /* TODO: figure out how to make this work */
-    if (flags & GA_BUFFER_HOST) FAIL(NULL, GA_DEVSUP_ERROR);
+static int allocate(cuda_context *ctx, gpudata **res, gpudata **prev,
+                    size_t size) {
+  CUdeviceptr ptr;
+  gpudata *next;
+  *prev = NULL;
 
-    res = malloc(sizeof(*res));
-    if (res == NULL) FAIL(NULL, GA_SYS_ERROR);
-    res->refcnt = 1;
+  if (size < BLOCK_SIZE) size = BLOCK_SIZE;
 
-    res->sz = size;
-    res->flags = flags & (GA_BUFFER_READ_ONLY|GA_BUFFER_WRITE_ONLY);
+  cuda_enter(ctx);
 
-    cuda_enter(ctx);
+  ctx->err = cuMemAlloc(&ptr, size);
+  if (ctx->err != CUDA_SUCCESS) {
+    cuda_exit(ctx);
+    return GA_IMPL_ERROR;
+  }
 
-    if (ctx->flags & GA_CTX_MULTI_THREAD)
-      fl |= CU_EVENT_BLOCKING_SYNC;
-    ctx->err = cuEventCreate(&res->ev, fl);
+  *res = new_gpudata(ctx, ptr, size);
 
-    if (ctx->err != CUDA_SUCCESS) {
-      free(res);
-      cuda_exit(ctx);
-      FAIL(NULL, GA_IMPL_ERROR);
-    }
+  cuda_exit(ctx);
 
-    if (size == 0) size = 1;
+  if (*res == NULL) {
+    cuMemFree(ptr);
+    return GA_MEMORY_ERROR;
+  }
 
-    ctx->err = cuMemAlloc(&res->ptr, size);
-    if (ctx->err != CUDA_SUCCESS) {
-        cuEventDestroy(res->ev);
-        free(res);
-        cuda_exit(ctx);
-        FAIL(NULL, GA_IMPL_ERROR);
-    }
-    res->ctx = ctx;
-    ctx->refcnt++;
+  (*res)->flags |= CUDA_HEAD_ALLOC;
 
-    if (flags & GA_BUFFER_INIT) {
-      ctx->err = cuMemcpyHtoD(res->ptr, data, size);
-      if (ctx->err != CUDA_SUCCESS) {
-	cuda_free(res);
-	FAIL(NULL, GA_IMPL_ERROR)
-      }
+  /* Now that the block is allocated, enter it in the freelist */
+  next = ctx->freeblocks;
+  for (; next && next->ptr < (*res)->ptr; next = next->next) {
+    *prev = next;
+  }
+  (*res)->next = next;
+  if (*prev)
+    (*prev)->next = *res;
+  else
+    ctx->freeblocks = *res;
+
+  return GA_NO_ERROR;
+}
+
+static int extract(gpudata *curr, gpudata *prev, size_t size) {
```

Comment:
add a comment that tell what this method do. It took me some times to understand that.